services/subscriber/service.go

```diff
@@ -0,0 +1,216 @@
+package subscriber
+
+import (
+	"expvar"
+	"fmt"
+	"log"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/influxdb/influxdb"
+	"github.com/influxdb/influxdb/cluster"
+	"github.com/influxdb/influxdb/meta"
+)
+
+// Statistics for the CQ service.
+const (
+	statPointsWritten = "points_written"
+	statWriteFailures = "write_failures"
+)
+
+type PointsWriter interface {
+	WritePoints(p *cluster.WritePointsRequest) error
+}
+
+// unique set that identifies a given subscription
+type subEntry struct {
+	db   string
+	rp   string
+	name string
+}
+
+type Service struct {
+	subs      map[subEntry]PointsWriter
+	MetaStore interface {
+		Databases() ([]meta.DatabaseInfo, error)
+	}
+	NewPointsWriter func(u url.URL) (PointsWriter, error)
+	Logger          *log.Logger
+	statMap         *expvar.Map
+	ticker          *time.Ticker
+}
+
+func NewService() *Service {
+	return &Service{
+		subs:            make(map[subEntry]PointsWriter),
+		NewPointsWriter: newPointsWriter,
+		Logger:          log.New(os.Stderr, "[subscriber] ", log.LstdFlags),
+		statMap:         influxdb.NewStatistics("subscriber", "subscriber", nil),
+	}
+}
+
+func (s *Service) Open() error {
+	if s.MetaStore == nil {
+		panic("no meta store")
+	}
+	s.Update()
+	s.ticker = time.NewTicker(time.Minute)
+	go s.pollForMetaUpdates()
+
+	return nil
+}
+
+func (s *Service) Close() error {
+	if s.ticker != nil {
+		s.ticker.Stop()
+	}
+	return nil
+}
+
+func (s *Service) pollForMetaUpdates() {
+	for range s.ticker.C {
+		s.Update()
+	}
+}
+
+// start new and stop deleted subscriptions.
+func (s *Service) Update() error {
+	dbis, err := s.MetaStore.Databases()
+	if err != nil {
+		return err
+	}
+	allEntries := make(map[subEntry]bool, 0)
+	// Add in new subscriptions
+	for _, dbi := range dbis {
+		for _, rpi := range dbi.RetentionPolicies {
+			for _, si := range rpi.Subscriptions {
+				se := subEntry{
```

Comment:
possible to pull everything in the `Subscriptions` loop into a function? Nice to avoid a bunch of code at 4 indent levels.